File: example/ios/QuantActionsSDK/Domain/Model/SubscriptionWithQuestionnaires.swift
/// Struct containing the information about a study and all of its questionnaires.
public struct SubscriptionWithQuestionnaires {
    /// the ``Cohort`` (a.k.a. study)
    public let cohort: Cohort

    /// the array of ``Questionnaire`` associated with the study
    public let questionnaires: [Questionnaire]

    /// the id of the subscription (a.k.a. participationID)
    public let subscriptionID: String

    /// the list of tap device ids associated with the subscription
    public let tapDeviceIDs: [String]

    /// the time to live of the premium features associated with the subscription
    public let premiumFeaturesTTL: Int
}

/// When subscribing to a cohort with a cohort ID or a subscription ID, the call will return a model
/// which contains all information related to the cohort including necessary
/// permissions and privacy policy.
public struct Cohort {
    /// Identification UUID fo the cohort
    public let id: String

    /// Human readable title
    public let name: String

    /// Privacy policy
    public let privacyPolicy: String

    /// It is `true` if the device is allowed to withdraw, if `false` only the cohort manager can withdraw the device
    public let canWidthdraw: Bool
}

/// Struct containing all information about a questionnaire that the user can fill in.
public struct Questionnaire {
    /// UUID of the questionnaire
    public let id: String

    /// Name of the questionnaire
    public let name: String

    /// Informal description of the questionnaire
    public let description: String

    /// Code - only used internally
    public let code: String

    /// Cohort ID to which this questionnaire is bound
    public let cohortID: String

    /// Body of the questionnaire in JSON format
    public let body: String
}

File: lib/src/domain/models/subscription/subscription_with_questionnaires.dart
import 'package:json_annotation/json_annotation.dart';

import '../../domain.dart';

part 'subscription_with_questionnaires.g.dart';

/// Class containing the information about a cohort and all of its questionnaires.
@JsonSerializable()
class SubscriptionWithQuestionnaires {
  /// The cohort associated with the subscription
  final Cohort cohort;

  /// the list of [Questionnaire]s associated with the cohort
  final List<Questionnaire> listOfQuestionnaires;

  /// the id of the subscription (a.k.a. participationID)
  final String subscriptionId;

  /// the list of device ids associated with the subscription
  final List<String> tapDeviceIds;

  /// the time to live of the premium features associated with the subscription
  final int premiumFeaturesTTL;

  /// Constructor
  SubscriptionWithQuestionnaires({
    required this.cohort,
    required this.listOfQuestionnaires,
    required this.subscriptionId,
    required this.tapDeviceIds,
    required this.premiumFeaturesTTL,
  });

  /// Create a [SubscriptionWithQuestionnaires] from a json
  factory SubscriptionWithQuestionnaires.fromJson(Map<String, dynamic> json) =>
      _$SubscriptionWithQuestionnairesFromJson(json);

  /// Convert the [SubscriptionWithQuestionnaires] to a json
  Map<String, dynamic> toJson() => _$SubscriptionWithQuestionnairesToJson(this);
}
